src/entities.rs

```diff
@@ -52,155 +52,237 @@ impl fmt::Debug for Entity {
     }
 }
 
+/// An iterator returning a sequence of Entity values from `Entities::reserve_entities`.
+pub struct ReserveEntitiesIterator<'a> {
+    // Metas, so we can recover the current generation for anything in the freelist.
+    meta: &'a [EntityMeta],
+
+    // Reserved IDs formerly in the freelist to hand out.
+    id_slice: &'a [u32],
+
+    // New Entity IDs to hand out, outside the range of meta.len().
+    id_range: core::ops::Range<u32>,
+}
+
+impl<'a> Iterator for ReserveEntitiesIterator<'a> {
+    type Item = Entity;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((&id, rest)) = self.id_slice.split_last() {
+            // We had previously reserved an item in the freelist.
+            self.id_slice = rest;
+            let entity = Entity {
+                generation: self.meta[id as usize].generation,
+                id,
+            };
+            Some(entity)
+        } else if let Some(id) = self.id_range.next() {
+            // The freelist ran out, so we reserved a new ID.
+            let entity = Entity { generation: 0, id };
+            Some(entity)
+        } else {
+            None
+        }
```

Comment:
Could this be more concise/elegant if `id_slice` was a `std::slice::Iter` and using `Option::or_else` rather than chained `if`s?

Reply:
Good call, I switched `id_slice` to `id_iter`. But I don't see `or_else` or `map_or_else` working out any better here.

Also I restored the old `contains` behavior (though now using `map_or`) and added a comment and unit test.